api/pkg/transformer/feast/bigtablestore/client.go

```diff
@@ -0,0 +1,102 @@
+package bigtablestore
+
+import (
+	"context"
+	"encoding/base64"
+
+	"cloud.google.com/go/bigtable"
+
+	feast "github.com/feast-dev/feast/sdk/go"
+	"github.com/gojek/merlin/pkg/transformer/spec"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+// BigTableClient provides the means of retrieving Feast features directly from
+// BigTable without going through Feast Serving.
+type BigTableClient struct {
+	encoder *Encoder
+	tables  map[string]*bigtable.Table
+}
+
+// NewClient instantiate new instance of BigTableClient
+func NewClient(bigtableStorage *spec.BigTableStorage, featureSpecs []*spec.FeatureTable, metadata []*spec.FeatureTableMetadata) (*BigTableClient, error) {
+	tables := make(map[string]*bigtable.Table)
+	client, err := newBigtableClient(bigtableStorage)
+	if err != nil {
+		return nil, err
+	}
+	for _, featureSpec := range featureSpecs {
+		btn := entityKeysToBigTable(featureSpec.Project, featureSpec.Entities)
+		if _, exists := tables[btn]; !exists {
+			tables[btn] = client.Open(btn)
+		}
+	}
+	registry := newCachedCodecRegistry(tables)
+	encoder, err := newEncoder(registry, featureSpecs, metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BigTableClient{
+		encoder: encoder,
+		tables:  tables,
+	}, nil
+}
+
+func newBigtableClient(storage *spec.BigTableStorage) (*bigtable.Client, error) {
+	opt := storage.Option
+	btConfig := bigtable.ClientConfig{AppProfile: storage.AppProfile}
+	var clientOpts []option.ClientOption
+	clientOpts = append(clientOpts, option.WithGRPCConnectionPool(int(opt.GrpcConnectionPool)))
+
+	keepAliveParams := keepalive.ClientParameters{}
+	if opt.KeepAliveInterval != nil {
+		keepAliveParams.Time = opt.KeepAliveInterval.AsDuration()
+	}
+	if opt.KeepAliveTimeout != nil {
+		keepAliveParams.Timeout = opt.KeepAliveTimeout.AsDuration()
+	}
+	clientOpts = append(clientOpts, option.WithGRPCDialOption(grpc.WithKeepaliveParams(keepAliveParams)))
+
+	if opt.CredentialJson != "" {
+		credsByte, err := base64.StdEncoding.DecodeString(opt.CredentialJson)
+		if err != nil {
+			return nil, err
+		}
+		clientOpts = append(clientOpts, option.WithCredentialsJSON(credsByte))
+	}
+	client, err := bigtable.NewClientWithConfig(context.Background(), storage.Project, storage.Instance, btConfig, clientOpts...)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// GetOnlineFeatures will first convert the request to RowQuery, which contains all
+// information needed to perform a query to BigTable to retrieve the Feast features,
+// stored in Avro format. The retrieved BigTable rows are then decoded back into
+// the Feast feature values again.
+func (b BigTableClient) GetOnlineFeatures(ctx context.Context, req *feast.OnlineFeaturesRequest) (*feast.OnlineFeaturesResponse, error) {
```

Comment:
The `client.go` is not tested. Can you add test for it? particularly for `GetOnlineFeatures`